internal/config/rule.go

```diff
@@ -1,195 +1,126 @@
+						return fmt.Errorf("updated rule refers unknown time period %d", update.TimePeriodID.Int64)
+					}
 
-		if escalation.ConditionExpr.Valid {
-			cond, err := filter.Parse(escalation.ConditionExpr.String)
-			if err != nil {
-				escalationLogger.Warnw("ignoring escalation, failed to parse condition", zap.Error(err))
-				continue
+					curElement.TimePeriod = timePeriod
+				} else {
+					curElement.TimePeriod = nil
+				}
 			}
 
-			escalation.Condition = cond
-		}
-
-		if escalation.FallbackForID.Valid {
-			// TODO: implement fallbacks (needs extra validation: mismatching rule_id, cycles)
-			escalationLogger.Warnw("ignoring fallback escalation (not yet implemented)")
-			continue
-		}
-
-		rule.Escalations[escalation.ID] = escalation
-		escalationsByID[escalation.ID] = escalation
-		escalationLogger.Debugw("loaded escalation config")
-	}
-
-	var recipientPtr *rule.EscalationRecipient
-	stmt = r.db.BuildSelectStmt(recipientPtr, recipientPtr)
-	r.logger.Debugf("Executing query %q", stmt)
-
-	var recipients []*rule.EscalationRecipient
-	if err := tx.SelectContext(ctx, &recipients, stmt); err != nil {
-		r.logger.Errorln(err)
-		return err
-	}
-
-	for _, recipient := range recipients {
-		recipientLogger := r.logger.With(
-			zap.Int64("id", recipient.ID),
-			zap.Int64("escalation_id", recipient.EscalationID),
-			zap.Int64("channel_id", recipient.ChannelID.Int64))
-
-		escalation := escalationsByID[recipient.EscalationID]
-		if escalation == nil {
-			recipientLogger.Warnw("ignoring recipient for unknown escalation")
-		} else {
-			escalation.Recipients = append(escalation.Recipients, recipient)
-			recipientLogger.Debugw("loaded escalation recipient config")
-		}
-	}
-
-	if r.Rules != nil {
-		// mark no longer existing rules for deletion
-		for id := range r.Rules {
-			if _, ok := rulesByID[id]; !ok {
-				rulesByID[id] = nil
+			return nil
+		},
+		nil)
+
+	incrementalApplyPending(
+		r,
+		&r.ruleEscalations, &r.configChange.ruleEscalations,
+		func(newElement *rule.Escalation) error {
+			elementRule, ok := r.Rules[newElement.RuleID]
+			if !ok {
+				return fmt.Errorf("rule escalation refers unknown rule %d", newElement.RuleID)
 			}
-		}
-	}
-
-	r.pending.Rules = rulesByID
-
-	return nil
-}
-
-func (r *RuntimeConfig) applyPendingRules() {
-	if r.Rules == nil {
-		r.Rules = make(map[int64]*rule.Rule)
-	}
 
-	for id, pendingRule := range r.pending.Rules {
-		if pendingRule == nil {
-			delete(r.Rules, id)
-		} else {
-			ruleLogger := r.logger.With(zap.Inline(pendingRule))
-
-			if pendingRule.TimePeriodID.Valid {
-				if p := r.TimePeriods[pendingRule.TimePeriodID.Int64]; p == nil {
-					ruleLogger.Warnw("ignoring rule with unknown timeperiod_id")
-					continue
-				} else {
-					pendingRule.TimePeriod = p
-				}
+			elementRule.Escalations[newElement.ID] = newElement
+			return nil
+		},
+		func(curElement, update *rule.Escalation) error {
+			curElement.NameRaw = update.NameRaw
+			curElement.Condition = update.Condition
```

Comment:
You're updating the parsed escalation condition but forgot to also update the actual filter string in `ConditionExpr`?

Reply:
https://github.com/Icinga/icinga-notifications/pull/191#discussion_r1660670544